### ISSUE TITLE
Color: From Hexcodes to Eyeballs

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4526]
Color: From Hexcodes to Eyeballs<br>
<br>
http://bit.ly/2QngH6T<br>
<br>
Logo
Created using Figma<br>
<br>
<div><article><h1>Color: From Hexcodes to Eyeballs</h1> <p></p> <p><img src="http://jamie-wong.com/images/color/Hero.png"></p> <p><em>This post is also available in Russian: <a href="https://habr.com/post/353582/">Цвет: от шестнадцатеричных кодов до глаза</a>, and Japanese: <a href="https://postd.cc/color/">色：ヘキサコードから眼球まで</a>.</em></p> <p>Why do we perceive <code>background-color: #9B51E0</code> as this particular purple?</p> <p><img src="http://jamie-wong.com/images/color/Purple.png"></p> <p>This is one of those questions where I thought I’d known the answer for a long time, but as I inspected my understanding, I realized there were pretty significant gaps.</p> <p>Through an exploration of electromagnetic radiation, optical biology, colorimetry, and display hardware, I hope to start filling in some of these gaps. If you want to skip ahead, here’s the lay of the land we’ll be covering:</p> <p><nav id="TableOfContents"></nav></p> <p>Otherwise, let’s start with the physics.</p> <h1 id="electromagnetic-radiation">Electromagnetic radiation</h1> <p>Radio waves, microwaves, infrared, visible light, ultraviolet, x-rays, and gamma rays are all forms of electromagnetic radiation. While these things all go by different names, these names really only label different ranges of wavelengths within the electromagnetic spectrum.</p> <figure><img src="http://jamie-wong.com/images/color/electromagneticSpectrum.png"><figcaption>The electromagnetic spectrum</figcaption></figure><p>The smallest unit of electromagnetic radiation is a photon. The energy contained within a photon is proportional to the frequency of its corresponding wave, with high energy photons corresponding with high frequency waves.</p> <p>To really understand color, we need to first understand radiation. Let’s take a closer look at the radiation of an incandescent light bulb.</p> <figure><img src="http://jamie-wong.com/images/color/incandescent.png"><figcaption>Photo by <a href="https://unsplash.com/photos/HfR0W6HW_Cw">Alex Iby</a></figcaption></figure><p>We might want to know how much energy the bulb is radiating. The <strong>radiant flux</strong> ($$\Phi_e$$) of an object is the total energy emitted per second, and is measured in Watts. The radiant flux of a 100W incandescent lightbulb is about 80W, with the remaining 20W being converted directly to non-radiated heat.</p> <p>If we want to know how much of that energy comes from each wavelength, we can look at the <strong>spectral flux</strong>. The spectral flux ($$\Phi_{e,\lambda}$$) of an object is radiant flux per unit wavelength, and is typically measured in Watts/nanometer.</p> <p>If we were to graph the spectral flux of our incandescent lightbulb as a function of wavelength, it might look something like this:</p> <figure><img src="http://jamie-wong.com/images/color/SpectralFlux1.png"></figure><p>The area under this curve will give the radiant flux. As an equation, $$\Phi_e = \int_0^\infty \Phi_{e,\lambda}(\lambda) d\lambda$$. In this case, the area under the curve will be about 80W.</p> <figure><img src="http://jamie-wong.com/images/color/SpectralFlux2.png"></figure><p>$$\Phi_{e}^{\text{bulb}} = \int_0^\infty \Phi_{e,\lambda}^\text{bulb}(\lambda) d\lambda = 80\text{W}$$</p> <p>Now you might’ve heard from eco-friendly campaigns that incandescent lightbulbs are brutally inefficient, and might be thinking to yourself, “well, 80% doesn’t seem so bad”.</p> <p>And it’s true — an incandescent lightbulb is a pretty efficient way to convert electricity into radiation. Unfortunately, it’s a terrible way to convert electricity into <em>human visible</em> radiation.</p> <h1 id="visible-light">Visible light</h1> <p>Visible light is the wavelength range of electromagnetic radiation from $$\lambda = 380\text{nm}$$ to $$\lambda = 750\text{nm}$$. On our graph of an incandescent bulb, that’s the shaded region below.</p> <figure><img src="http://jamie-wong.com/images/color/SpectralFlux3.png"></figure><p>$$\int_{380 \text{nm}}^{750 \text{nm}} \Phi_{e,\lambda}^\text{bulb}(\lambda) d\lambda = 8.7W$$</p> <p>Okay, so the energy radiated <em>within the visible spectrum</em> is $$8.7W$$ for an efficiency of $$8.7\%$$. That seems pretty awful. But it gets worse.</p> <p>To understand why, let’s consider why visible light is, well, <em>visible</em>.</p> <h1 id="human-perceived-brightness">Human perceived brightness</h1> <figure><img src="http://jamie-wong.com/images/color/bweye.png"><figcaption>Photo by <a href="https://unsplash.com/photos/QaGNhezu_5Q">Christopher Burns</a></figcaption></figure><p>Just as we saw that an incandescent light bulb doesn’t radiate equally at all wavelengths, our eyes aren’t equally sensitive to radiation at all wavelengths. If we measure a human eye’s sensitivity to every wavelength, we get a <a href="https://en.wikipedia.org/wiki/Luminosity_function">luminosity function</a>. The standard luminosity function, $$\bar y(\lambda)$$ looks like this:</p> <figure><img src="http://jamie-wong.com/images/color/SpectralFlux4.png"></figure><p>The bounds of this luminosity function <em>define</em> the range of visible light. Anything outside this range isn’t visible because, well, our eyes aren’t sensitive to it!</p> <p>This curve also shows that our eyes are <em>much</em> more sensitive to radiation at 550nm than they are to radiation at either 650nm or 450nm.</p> <p>Other animals have eyes that are sensitive to a different range of wavelengths, and therefore different luminosity functions. Birds can see ultraviolet radiation in the range between $$\lambda=300\text{nm}$$ to $$\lambda=400\text{nm}$$, so if scholarly birds had defined the electromagnetic spectrum, that would’ve been part of the “visible light” range for them!</p> <figure><img src="http://jamie-wong.com/images/color/owl.png"><figcaption>Photo by <a href="https://unsplash.com/photos/0J6cTw0V2lE">Timothy Rhyne</a></figcaption></figure><p>By multiplying the graph of spectral flux with the luminosity function $$\bar y(\lambda)$$, we get a function which describes the contributions to human perceived brightness for each wavelength emitted by a light source.</p> <figure><img src="http://jamie-wong.com/images/color/SpectralFlux5.png"></figure><p>This is the <strong>spectral luminous flux (</strong>$$\Phi_{v,\lambda}$$<strong>)</strong>. To acknowledge this is about human perception rather than objective power, the luminous flux is typically measured in lumens rather than Watts using a conversion ratio of 683.002 lm/W.</p> <p>$$\Phi_{v,\lambda}(\lambda) = 683.002 \frac{\text{lm}}{\text{W}} \cdot \bar y(\lambda) \cdot \Phi_{e,\lambda}(\lambda)$$</p> <p>The <strong>luminous flux (</strong>$$\Phi_v$$<strong>)</strong> of a light source is the total <em>human perceived</em> power of the light.</p> <p>Just as we calculated the radiant flux by taking the area under the spectral flux curve, we can find the luminous flux by taking the area under the spectral <em>luminous</em> flux curve, with a constant conversion from perceived watts to lumens:</p> <figure><img src="http://jamie-wong.com/images/color/SpectralFlux5.5.png"></figure><p>$$\Phi_{v}^\text{bulb} = \int_0^\infty \bar y(\lambda) \cdot \Phi_{e,\lambda}^\text{bulb}(\lambda) d\lambda = 683.002 \frac{\text{lm}}{\text{W}} \cdot 2.4\text{W} \approx 1600 \text{lm}$$</p> <p>So the luminous flux of our 100W incandescent lightbulb is a measly 2.4W or 1600lm! The bulb has a luminous efficiency of 2.4%, a far cry from the 80% efficiency converting electricity into radiation.</p> <p>Perhaps if we had a light source that concentrated its emission into the visible range, we’d be able to get more efficient lighting. Let’s compare the spectra of incandescent, fluorescent, and LED bulbs:</p> <figure><img src="http://jamie-wong.com/images/color/SpectralFlux6.png"></figure><p>And indeed, we can see that far less of the radiation in fluorescent or led bulbs is wasted on wavelengths that humans can’t see. Where incandescent bulbs might have an efficiency of 1-3%, fluorescent bulbs can be around 10% efficient, and LED bulbs can achieve up to 20% efficiency!</p> <p>Enough about brightness, let’s return to the focus of this post: color!</p> <h1 id="quantifying-color">Quantifying color</h1> <figure><img src="http://jamie-wong.com/images/color/lemon.png"><figcaption>Photo by <a href="https://unsplash.com/photos/sil2Hx4iupI">Lauren Mancke</a></figcaption></figure><p>How might we identify a given color? If I have a lemon in front of me, how can I tell you over the phone what color it is? I might tell you “the lemon is yellow”, but which yellow? How would you precisely identify each of these yellows?</p> <p><img src="https://d2mxuefqeaa7sj.cloudfront.net/s_382135F512C449943D36A9C35B9E8A4F38EB382A3D72DF63DC63D011CB7A8322_1521414272066_Yellows.png"></p> <p>Armed with the knowledge that color is humans’ interpretation of electromagnetic radiation, we might be tempted to define color mathematically via spectral flux. Any human visible color will be some weighted combination of the monochromatic (single wavelength) colors. Monochromatic colors are also known as spectral colors.</p> <figure><img src="http://jamie-wong.com/images/color/Rainbow.png"><figcaption>The monochromatic colors by wavelength</figcaption></figure><p>For any given object, we can measure its emission (or reflectance) spectrum, and use that to precisely identify a color. If we can reproduce the spectrum, we can certainly reproduce the color!</p> <p>The sunlight reflected from a point on a lemon might have a reflectance spectrum that looks like this:</p> <figure><img src="http://jamie-wong.com/images/color/ReflectanceSpectrum.png"></figure><p><em>Note: the power and spectral distribution of radiation that reaches your eye is going to depend upon</em> <em>the</em> <em>power &amp; emission spectrum of the light source, the distance of the light source from the illuminated object, the size and shape of the object, the absorption spectrum of the object, and your distance from the object. That’s a lot to think about, so let’s focus just on what happens when that light hits your eye. Let’s also disregard units for now to focus on concepts.</em></p> <p>When energy with this spectral distribution hits our eyes, we perceive it as “yellow”. Now let’s say I take a photo of the lemon and upload it to my computer. Next, I carefully adjust the colors on my screen until a particular point of the on-screen lemon is imperceptibly different from the color of the actual lemon in my actual hand.</p> <p>If you were to measure the spectral power distribution coming off of the screen, what would you expect the distribution to look like? You might reasonably expect it to look similar to the reflectance spectrum of the lemon above. But it would actually look something like this:</p> <figure><img src="http://jamie-wong.com/images/color/EmissionSpectrum.png"></figure><p>Two different spectral power distributions that look the same to human observers are called <a href="https://en.wikipedia.org/wiki/Metamerism_(color)"><strong>metamers</strong></a>.</p> <figure><img src="http://jamie-wong.com/images/color/Metamers1.png"></figure><p>To understand how this is possible, let’s take a look at the biology of the eye.</p> <h1 id="optical-biology">Optical biology</h1> <figure><img src="http://jamie-wong.com/images/color/coloreye.png"><figcaption>Photo by <a href="https://unsplash.com/photos/UbJMy92p8wk">Amanda Dalbjörn </a></figcaption></figure><p>Our perception of light is the responsibility of specialized cells in our eyes called “rods” and “cones”. Rods are predominately important in low-light settings, and don’t play much role in color vision, so we’ll focus on the cones.</p> <p>Humans typically have 3 kinds of cones. Having three different kinds of cones makes humans “trichromats”. There is, however, at least one confirmed case of a <a href="http://nymag.com/scienceofus/2015/02/what-like-see-a-hundred-million-colors.html">tetrochromat human</a>! Other animals have even more cone categories. <a href="http://theoatmeal.com/comics/mantis_shrimp">Mantis shrimp</a> have <em>sixteen</em> different kinds of cones.</p> <p>Each kind of cone is labelled by the range of wavelengths of light they are excited by. The standard labelling is “S”, “M”, and “L” (short, medium, long).</p> <figure><img src="http://jamie-wong.com/images/color/Cones.png"></figure><p>These three curves indicate how sensitive the corresponding cone is to each wavelength. The highest point on each curve is called the “peak wavelength”, indicating the wavelength of radiation that the cone is most sensitive to.</p> <p>Let’s see how our cones process the light bouncing off the lemon in my hand.</p> <figure><img src="http://jamie-wong.com/images/color/ConeExcitation1.png"></figure><p>By looking at the normalized areas under the curves, we can see how much the radiation reflected from the real lemon excites each of cones. In this case, the normalized excitations of the S, M, and L cones are 0.02, 0.12, and 0.16 respectively. Now let’s repeat the process for the on-screen lemon.</p> <figure><img src="http://jamie-wong.com/images/color/ConeExcitation2.png"></figure><p>Despite having totally different radiation spectra reaching the eye, the cone <em>excitations</em> are the same (S=0.02, M=0.12, L=0.16). That’s why the point on the real lemon and the point on the digital lemon look the same to us!</p> <figure><img src="http://jamie-wong.com/images/color/Metamers2.png"><figcaption> The normalized cone area under the stimulation curves will always be equal for all 3 cone types in the case of metamers. </figcaption></figure><p>Our 3 sets of cones reduce any spectral flux curve $$\Phi_e(\lambda)$$ down to a triplet of three numbers $$(S, M, L)$$, and every distinct $$(S, M, L)$$ triplet will be a distinct color! This is pretty convenient, because (0.02, 0.12, 0.16) is much easier to communicate than a complicated continuous function. For the mathematically inclined, our eyes are doing a dimensional reduction from an infinite dimensional space into 3 dimensions, which is a pretty damn cool thing to be able to do subconsciously.</p> <p>This $$(S, M, L)$$ triplet is, in fact, our first example of a <strong>color space</strong>.</p> <h1 id="color-spaces">Color spaces</h1> <p>Color spaces allow us to define with numeric precision what color we’re talking about. In the previous section, we saw that a specific yellow could be represented as (0.02, 0.12, 0.16) in the SML color space, which is more commonly known as the <a href="https://en.wikipedia.org/wiki/LMS_color_space">LMS color space</a>.</p> <p>Since this color space is describing the stimulation of cones, by definition any human visible color can be represented by positive LMS coordinates (excluding the extremely rare tetrachromat humans, who would need four coordinates instead of three).</p> <p>But, alas, this color space has some unhelpful properties.</p> <p>For one, not all triplet values (also called <strong>tristimulus values)</strong> are <em>physically possible</em>. Consider the LMS coordinates (0, 1, 0). To physically achieve this coordinate, we would need to find some way of stimulating the M cones without stimulating the L or S cones <em>at all</em>. Because the M cone’s sensitivity curve significantly overlaps at least one of L or S at all wavelengths, this is impossible!</p> <figure><img src="http://jamie-wong.com/images/color/Cones.png"><figcaption> Any wavelength which stimulates the M cone will also stimulate either the L or S cone (or both!) </figcaption></figure><p>A problematic side effect of this fact is that it’s really difficult to increase stimulation of only one of the cones. This, in particular, would make it not a great candidate for building display hardware.</p> <p>Another historical, pragmatic problem was that the cone sensitivities weren’t accurately known until the 1990’s, and a need to develop a mathematically precise model of color significantly predates that. The first significant progress on that front came about in the late 1920’s.</p> <h1 id="wright-guild-s-color-matching-experiments">Wright &amp; Guild’s color matching experiments</h1> <p>In the late 1920’s, William David Wright and John Guild conducted experiments to precisely define color in terms of contributions from 3 specific wavelengths of light.</p> <p>Even though they may not have known about the three classes of cones in the eye, the idea that all visible colors could be created as the combination of three colors had been proposed at least a hundred years earlier.</p> <figure><img src="http://jamie-wong.com/images/color/tricolor.png"><figcaption> An example of a tricolor theory by Charles Hayter, 1826 </figcaption></figure><p>Wright &amp; Guild had the idea to construct an apparatus that would allow test subjects to reconstruct a test color as the combination of three fixed wavelength light sources. The setup would’ve looked something like this:</p> <figure><img src="http://jamie-wong.com/images/color/ColorMatching1.png"></figure><p>The experimenter would set the lamp on the bottom to a target wavelength, (for instance, 600nm) then ask the test subject to adjust the three lamp power controls until the colors they were seeing matched.</p> <figure><img src="http://jamie-wong.com/images/color/ColorMatching2.png"></figure><p>The power settings of the three dials give us a (red, green, blue) triplet identifying the pure spectral color associated with 600nm. Repeating this process every 5nm with about 10 test subjects, a graph emerges showing the amounts of red (700nm), green (546nm), and blue (435nm) light needed to reconstruct the appearance of a given wavelength. These functions are known as <strong>color matching function (CMFs)</strong>.</p> <p>These particular color matching functions are known as $$\bar r(\lambda)$$, $$\bar g(\lambda)$$, and $$\bar b (\lambda)$$.</p> <figure><img src="http://jamie-wong.com/images/color/cmfs1.png"></figure><p>This gives the pure spectral color associated with 600nm an $$(R, G, B)$$ coordinate of (0.34, 0.062, 0.00). This is a value in the <a href="https://en.wikipedia.org/wiki/CIE_1931_color_space#CIE_RGB_color_space">CIE 1931 RGB color space</a>.</p> <p>Hold on though — what does it mean when the functions go negative, like here?</p> <figure><img src="http://jamie-wong.com/images/color/cmfs2.png"></figure><p>The pure spectral color associated with 500nm has an $$(R, G, B)$$ coordinate of (-0.72, 0.85, 0.48). So what exactly does that -0.72 mean?</p> <p>It turns out that no matter what you set the red (700nm) dial to, it will be impossible to match a light outputting at 500nm, regardless of the values of blue and green dials. You can, however, make the two sides match by adding red light to the <em>bottom</em> side.</p> <figure><img src="http://jamie-wong.com/images/color/ColorMatching3.png"></figure><p>The actual setup probably would’ve had a full set of 3 variable power, fixed wavelength lights on either side of the divider to allow any of them to be adjusted to go negative.</p> <p>Using our color matching functions, we can match any monochromatic light using a combination of (possibly negative) amounts of red (700nm), green (546nm), and blue (435nm) light.</p> <p>Just as we were able to use our L, M, and S cone sensitivity functions to determine cone excitation for any spectral distribution, we can do the same thing with our color matching functions. Let’s apply that to the lemon color from before:</p> <figure><img src="http://jamie-wong.com/images/color/ColorMatchingLemon.png"></figure><p>By taking the area under the curve of the product of the spectral curve and the color matching functions, we’re left with an $$(R, G, B)$$ triplet (1.0, 0.8, 0.2) uniquely identifying this color.</p> <p>While the $$(L, M, S)$$ color space gave us a precise way to <em>identify</em> colors, this $$(R, G, B)$$ color space gives us a precise way to <em>reproduce</em> colors. But, as we saw in the color matching functions, any colors with a negative $$(R, G, B)$$ coordinate can’t actually be reproduced.</p> <figure><img src="http://jamie-wong.com/images/color/cmfs3.png"></figure><p>But this graph only shows which spectral colors can’t be reproduced. What about non-spectral colors? Can I produce pink with an R, G, B combination? What about teal?</p> <p>To answer these questions, we’ll need a better way of visualizing color space.</p> <h1 id="visualizing-color-spaces-chromaticity">Visualizing color spaces &amp; chromaticity</h1> <p>So far most of our graphs have put wavelength on the horizontal axis, and we’ve plotted multiple series to represent the other values of interest.</p> <figure><img src="http://jamie-wong.com/images/color/WavelengthPlots.png"></figure><p>Instead, we could plot color as a function of $$(R, G, B)$$ or $$(L, M, S)$$. Let’s see what color plotted in 3D $$(R, G, B)$$ space looks like.</p> <figure><img src="http://jamie-wong.com/images/color/LinearRGBCube.png"></figure><p>Cool! This gives us a visualization of a broader set of colors, not just the spectral colors of the rainbow.</p> <p>A simple way to reduce this down to two dimensions would be to have a separate plot for each pair of values, like so:</p> <figure><img src="http://jamie-wong.com/images/color/RGBPairPlots.png"><figcaption>Component pairs plotted, holding the third coordinate constant</figcaption></figure><p>In each of these plots, we discard one dimension by holding one thing constant. Rather than holding one of red, green, and blue constant, it would be really nice to have a plot showing all the colors of the rainbow &amp; their combinations, while holding <em>lightness</em> constant.</p> <p>Looking at the cube pictures again, we can see that (0, 0, 0) is black, and (1, 1, 1) is white.</p> <figure><img src="http://jamie-wong.com/images/color/LinearRGBCube.png"></figure><p>What happens if we slice the cube diagonally across the plane containing $$(1, 0, 0)$$, $$(0, 1, 0)$$, and $$(0, 0, 1)$$?</p> <figure><img src="http://jamie-wong.com/images/color/TriangleSliceRGB.png"></figure><p>This triangle slice of the cube has the property that $$R + G + B = 1$$, and we can use $$R + G + B$$ as a crude approximation of lightness. If we take a top-down view of this triangular slice, then we get this:</p> <figure><img src="http://jamie-wong.com/images/color/rgChromaticity1.png"></figure><p>This two dimensional representation of color is called <strong>chromaticity</strong>. This particular kind is called <a href="https://en.wikipedia.org/wiki/Rg_chromaticity"><strong>rg chromaticity</strong></a>. Chromaticity gives us information about the <em>ratio</em> of the primary colors independent of the lightness.</p> <p>This means we can have the same chromaticity at many different intensities.</p> <figure><img src="http://jamie-wong.com/images/color/rgChromaticity2.png"></figure><p>We can even make a chromaticity graph where the intensity varies with r &amp; g in order to maximize intensity while preserving the ratio between $$R$$, $$G$$, and $$B$$.</p> <figure><img src="http://jamie-wong.com/images/color/rgChromaticity3.png"></figure><p>Chromaticity is a useful property of a color to consider because it stays constant as the intensity of a light source changes, so long as the light source retains the same spectral distribution. As you change the brightness of your screen, chromaticity is the thing that stays constant!</p> <p>There are many different ways of dividing chromaticity into two dimensions. One of the common methods is used in both the HSL and HSV color spaces. Both color spaces split chromaticity into “hue” and “saturation”, like so:</p> <figure><img src="http://jamie-wong.com/images/color/HSL.png"></figure><p>It might appear at a glance that the rg chromaticity triangle and these hue vs. saturation squares contains every color of the rainbow. It’s time to revisit those pesky negative values in our color matching functions.</p> <h1 id="gamuts-and-the-spectral-locus">Gamuts and the spectral locus</h1> <p>If we take our color matching functions $$\bar r(\lambda)$$, $$\bar g(\lambda)$$, and $$\bar b(\lambda)$$ and use them to plot the rg chromaticities of the spectral colors, we end up with a plot like this:</p> <figure><img src="http://jamie-wong.com/images/color/rgChromaticityPlot1.png"></figure><p>The black curve with the colorful dots on it shows the chromaticities of all the pure spectral colors. The curve is called the <strong>spectral locus</strong>. The stars mark the wavelengths of the variable power test lamps used in the color matching experiments.</p> <p>If we overlay our previous chromaticity triangles onto this chart, we’re left with this:</p> <figure><img src="http://jamie-wong.com/images/color/rgChromaticityPlot2.png"></figure><p>The area inside the spectral locus represents all of the chromaticities that are visible to humans. The checkerboard area represents chromaticities that humans can recognize, but that <em>cannot</em> be reproduced by any positive combination of 435nm, 546nm, and 700nm lights. From this diagram, we can see that we’re unable to reproduce any of the spectral colors between 435nm and 546 nm, which includes pure cyan.</p> <p>The triangle on the right without the checkerboard is all of the chromaticities that <em>can</em> be reproduced by a positive combination. We call the area that can be reproduced the <strong>gamut</strong> of the color space.</p> <p>Before we can <em>finally</em> return to hexcodes, we have one more color space we need to cover.</p> <h1 id="cie-xyz-color-space">CIE XYZ color space</h1> <p>In 1931, the International Comission on Illumination convened and created two color spaces. The first was the RGB color space we’ve already discussed, which was created based on the results of Wright &amp; Guild’s color matching experiments. The second was the XYZ color space.</p> <p>One of the goals of the XYZ color space was to have positive values for all human visible colors, and therefore have all chromaticities fit in the range [0, 1] on both axes. To achieve this, a linear transformation of RGB space was carefully selected.</p> <p> $$ \begin{bmatrix} X \\ Y \\ Z \end{bmatrix} = \frac{1}{b_{21}} \begin{bmatrix} b_{11} &amp; b_{12} &amp; b_{13} \\ b_{21} &amp; b_{22} &amp; b_{23} \\ b_{31} &amp; b_{32} &amp; b_{33} \end{bmatrix} \begin{bmatrix} R \\ G \\ B \end{bmatrix} = \frac{1}{0.17697} \begin{bmatrix} 0.49000 &amp; 0.31000 &amp; 0.20000 \\ 0.17697 &amp; 0.81240 &amp; 0.010630 \\ 0.0000 &amp; 0.010000 &amp; 0.99000 \end{bmatrix} \begin{bmatrix} R \\ G \\ B \end{bmatrix} $$</p> <p>The analog of rg chromaticity for XYZ space is xy chromaticity and is the more standard coordinate system used for chromaticities diagrams.</p> <figure><img src="http://jamie-wong.com/images/color/xyChromaticityPlot.png"></figure><p>Gamuts are typically represented by a triangle placed into an xy chromaticity diagram. For instance, here’s the gamut of CIE RGB again, this time in xy space.</p> <figure><img src="http://jamie-wong.com/images/color/gamut1.png"></figure><p>With an understanding of gamuts &amp; chromaticity, we can finally start to discuss how digital displays are able to display an intended color.</p> <h1 id="screen-subpixels">Screen subpixels</h1> <p>Regardless of the manufacturer of your display, if you took a powerful magnifying glass to your display, you would find a grid of pixels, where each pixel is composed of 3 types of subpixels: one type emitting red, one green, and one blue. It might look something like this:</p> <figure><img src="http://jamie-wong.com/images/color/Subpixels.png"></figure><p>Unlike the test lamps used in the color matching experiments, the subpixels do not emit monochromatic light. Each type of subpixel has its own spectral distribution, and these will vary from device to device.</p> <figure><img src="http://jamie-wong.com/images/color/subpixelSpectra.png"><figcaption>MacBook Air subpixel spectral data from <a href="https://fluxometer.com/rainbow/">f.luxometer</a></figcaption></figure><p>Using <a href="https://support.apple.com/guide/colorsync-utility/welcome/mac">ColorSync Utility</a> on my Macbook Pro, I was able to determine the xy space gamut of my screen.</p> <figure><img src="http://jamie-wong.com/images/color/gamut3.png"></figure><p>Notice that the corners of the gamut no longer lie along the spectral locus. This makes sense, since the subpixels do not emit pure monochromatic light. This gamut represents the full range of chromaticities that this monitor can faithfully reproduce.</p> <p>While gamuts of monitors will vary, modern monitors should try to enclose a specific other gamut: sRGB.</p> <h1 id="srgb">sRGB</h1> <p>sRGB (“standard Red Green Blue”) is a color space created by HP and Microsoft in 1996 to help ensure that color data was being transferred faithfully between mediums.</p> <p>The standard specifies the chromaticities of the red, green, and blue primaries.</p> <table>
<thead><tr>
<th>Chromaticity</th> <th>Red</th> <th>Green</th> <th>Blue</th>
</tr></thead>
<tbody>
<tr>
<td>x</td> <td>0.6400</td> <td>0.3000</td> <td>0.1500</td>
</tr>
<tr>
<td>y</td> <td>0.3300</td> <td>0.6000</td> <td>0.0600</td>
</tr>
<tr>
<td>Y</td> <td>0.2126</td> <td>0.751</td> <td>0.0722</td>
</tr>
</tbody>
</table>
<p>If we plot these, we wind up with a gamut similar to, but slightly smaller than, the MacBook LCD screen.</p> <figure><img src="http://jamie-wong.com/images/color/gamut2.png"></figure><p>There are parts of the official sRGB gamut that aren’t within the MacBook Pro LCD gamut, meaning that the LCD can’t faithfully reproduce them. To accommodate for that, my MacBook seems to use a modified sRGB gamut.</p> <figure><img src="http://jamie-wong.com/images/color/gamut4.png"></figure><p>sRGB is the default color space used almost everywhere, and is the standard color space used by browsers (<a href="https://www.w3.org/TR/css-color-4/#color-type">specified in the CSS standard</a>). All of the diagrams in this blog post are in sRGB color space. That means that all colors outside of the sRGB gamut aren’t accurately reproduced in the diagrams in this post!</p> <p>Which brings us, finally, to how colors are specified on the web.</p> <h1 id="srgb-hexcodes">sRGB hexcodes</h1> <p><code>#9B51E0</code> specifies a color in sRGB space. To convert it to its associated (R, G, B) coordinate, we divide each of the three components by <code>0xFF</code> aka 255. In this case:</p> <pre><code>0x9B/0xFF = 0.61 0x51/0xFF = 0.32 0xE0/0xFF = 0.88 </code></pre> <p>So the coordinate associated with <code>#9BE1E0</code> is $$(0.61, 0.32, 0.88)$$.</p> <p>Before we send these values to the display hardware to set subpixel intensities, there’s <em>one</em> more step: gamma correction.</p> <h1 id="gamma-correction">Gamma correction</h1> <p>With each coordinate in RGB space being given 256 possible values, we want to ensure that each adjacent pair is as different as possible. For example, we want <code>#030000</code> to be as different from <code>#040000</code> as <code>#F40000</code> is from <code>#F50000</code>.</p> <p>Human vision is much more sensitive to small changes in low energy lights than small changes to high energy lights, so we want to allocate more of the 256 values to representing low energy values.</p> <p>To see how, let’s imagine we wanted to encode greyscale values, and only had 3 bits to do it, giving us 8 possible values.</p> <p>If we plot grey values as a linear function of energy, it would look something like this:</p> <figure><img src="http://jamie-wong.com/images/color/linearEnergy.png"></figure><p>We’ll call our 3 bit encoded value $$Y$$. If our encoding scheme spaces out each value we encode evenly ($$Y = \frac{\left\lfloor8E\right\rfloor}{8}$$), then it would look like this:</p> <figure><img src="http://jamie-wong.com/images/color/gamma1.png"></figure><p>You can see that the perceptual difference between $$Y=0$$ and $$Y=1$$ is significantly greater than the difference between $$Y=6$$ and $$Y=7$$.</p> <p>Now let’s see what happens if we use a power function instead. Let’s try $$Y = \left(\frac{\left\lfloor8E\right\rfloor}{8}\right)^2$$.</p> <figure><img src="http://jamie-wong.com/images/color/gamma2.png"></figure><p>We’re getting much closer to perceptual uniformity here, where each adjacent pair of values is as different as any other adjacent pair.</p> <p>This process of taking energy values and mapping them to discrete values is called <strong>gamma encoding</strong>. The inverse operations (converting discrete values to energy values) is called <strong>gamma decoding</strong>.</p> <p>In general form, gamma correction has the equation $$V_{out} = A V_{in}^\gamma$$. The exponent is the greek letter “gamma”, hence the name.</p> <p>The encoding &amp; decoding rules for sRGB use a similar idea, but slightly more complex.</p> <p>$$C_\mathrm{linear}= \begin{cases}\frac{C_\mathrm{sRGB}}{12.92}, &amp; C_\mathrm{sRGB}\le0.04045\\ \left(\frac{C_\mathrm{sRGB}+0.055}{1.055}\right)^{2.4}, &amp; C_\mathrm{sRGB}&gt;0.04045 \end{cases}$$</p> <p>If we plot sRGB values against linear values, it would look like this:</p> <figure><img src="http://jamie-wong.com/images/color/gamma3.png"></figure><p>Okay! That was the last piece we needed to understand to see how we get from hex codes to eyeballs! Let’s do the walkthrough 😀</p> <h1 id="from-hexcodes-to-eyeballs">From Hexcodes to Eyeballs</h1> <p>First, we take <code>#9B51E0</code>, split it up into its R, G, B components, and normalize those components to be the range $$[0, 1]$$.</p> <figure><img src="http://jamie-wong.com/images/color/summary1.png"></figure><p>This gives us a coordinate of $$(0.61, 0.32, 0.88)$$ in sRGB space. Next, we take our sRGB components and convert them to linear values.</p> <figure><img src="http://jamie-wong.com/images/color/summary2.png"></figure><p>This gives us a coordinate $$(0.33, 0.08, 0.10)$$ in linear RGB space. These values are used to set the intensity of the subpixels on the screen.</p> <figure><img src="http://jamie-wong.com/images/color/summary3.png"></figure><p>The spectral distributions of the subpixels combine to a single spectral distribution for the whole pixel.</p> <figure><img src="http://jamie-wong.com/images/color/summary4.png"></figure><p>The electromagnetic radiation travels from the pixel through your cornea and hits your retina, exciting your 3 kinds of cones.</p> <figure><img src="http://jamie-wong.com/images/color/summary5.png"></figure><p>Putting it all together for a different color, we’re left with the image that opens this post!</p> <figure><img src="http://jamie-wong.com/images/color/Hero.png"></figure><h1 id="a-brief-note-about-brightness-setting">A brief note about brightness setting</h1> <figure><img src="http://jamie-wong.com/images/color/brightness.png"></figure><p>Before sRGB values are converted into subpixel brightness, they’ll be attenuated by the device’s brightness setting. So the <code>0xFF0000</code> on a display at 50% brightness might match the <code>0x7F0000</code> on the same display at 100% brightness.</p> <p>In an ideal screen, this would mean that regardless of the brightness setting, black pixels $$(0, 0, 0)$$ would emit no light. Most phone &amp; laptop screens are LCD screens, however, where each subpixel is a filter acting upon white light. This video is a great teardown of how LCDs work:</p> VIDEO <p>The filter is imperfect, so as brightness is increased, black pixels will emit light as the backlight bleeds through. OLED screens (like on the iPhone X and Pixel 2) don’t use a backlight, allowing them to have a consistent black independent of screen brightness.</p> <h1 id="stuff-i-left-out">Stuff I left out</h1> <p>This post intentionally glosses over many facets of color reproduction and recognition. For instance, we didn’t talk about what your brain does with the cone excitation information in the <a href="https://psych.ucalgary.ca/PACE/VA-Lab/colourperceptionweb/theories.htm">opponent-process theory</a> or the effects of <a href="https://en.wikipedia.org/wiki/Color_constancy">color constancy</a>. We didn’t talk about <a href="https://en.wikipedia.org/wiki/Additive_color">additive color</a> vs. <a href="https://en.wikipedia.org/wiki/Subtractive_color">subtractive color</a>. We didn’t talk about <a href="http://www.colour-blindness.com/general/how-it-works-science/">color blindness</a>. We didn’t talk about the difference between <a href="https://en.wikipedia.org/wiki/Photometry_(optics)#Photometric_quantities">luminous flux, luminous intensity, luminance, illuminance, and luminous emittance</a>. We didn’t talk about <a href="https://en.wikipedia.org/wiki/ICC_profile">ICC device color profiles</a> or what programs like <a href="https://justgetflux.com/">f.lux</a> do to color perception.</p> <p>I left them out because this post is already way too long! As a <a href="https://twitter.com/amtinits">friend of mine</a> said: even if you’re a person who understands that most things are deeper than they look, color is way deeper than you would reasonably expect.</p> <h1 id="references">References</h1> <p>I spent an unusually large portion of the time writing this post just reading because I kept discovering that I was missing something I needed to explain as completely as I’d like.</p> <p>Here’s a short list of the more helpful ones:</p> <p>I also needed to draw upon many data tables to produce the charts in this post:</p> <p>Special thanks to <a href="http://www.coopernetics.com/blog/">Chris Cooper</a> and <a href="http://rykap.com/">Ryan Kaplan</a> for providing feedback on the draft of this post.</p> </article></div><br>
<br>
<br>
<br>
<br>
<br>
via Zero Wind :: Jamie Wong http://jamie-wong.com<br>
<br>
June 22, 2019 at 12:36PM
